### ISSUE TITLE
Need control over the always-on call stack

### ISSUE THREAD

**Issue by Rolemancer** [NLog/NLog#3914]
**NLog version**: 4.7
**NLog version**: .Net 4.5

The idea of logs is predictable and controllable data. However, current solution of `LoggerImpl.Write()` always saves callstack. What if not all log records need this, but we need configureate `{stacktrace}`?. All I could find about this question is this: [stackoverflow](https://stackoverflow.com/questions/53119784/customizing-when-to-capture-stack-trace-in-nlog)

That's about what I'd like to get:
Config file where I can configure will stack trace always enabled or not. Second thing - layout config for stack trace like for exceptions. Config shouldn't mean, that stack trace should always be enabled.

```xml
<nlog throwConfigExceptions="true" throwExceptions="true" xmlns="http://www.nlog-project.org/schemas/NLog.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <targets async="true">
    <target name="xmlFile" xsi:type="File" fileName="E:\log.xml" maxArchiveFiles="3" archiveNumbering="Sequence" archiveDateFormat="dd-mm-yyyy" archiveOldFileOnStartup="true" stacktraceAlwaysEnabled="false">
      <layout xsi:type="XmlLayout" indentXml="true" includeAllProperties="true" maxRecursionLimit="10">
        <element name="message" value="${message}" />
        <element name="exception" value="${exception:format=toString}" />
        <element name="stacktrace" value="${custom.stacktrace:format=Raw:separator=&#13;&#10;}" /> 
      </layout>
    </target>
  </targets>
  <rules>
    <logger name="*" minlevel="Debug" writeTo="xmlFile" />
  </rules>
</nlog>

```
Next two messeges should give different logs:
First
```C#
Log.Debug()
      .Message("Hello")
      .Write();
```
Would give
```xml
<logevent>
  <message>Hello</message>
</logevent>
```

Second:  
```C#
 Log.Debug()
       .Message("Hello")
       .StackTrace(new StackTrace(), 0)
       .Write();
```
Would give
```xml
<logevent>
  <message>Hello</message>
  <stacktrace>some stack </stacktrace>
</logevent>
```
The behaviour make be the same like with exceptions.

I didn't find how to configurate same behaviour right now.

**Comment by snakefoot** [comment]
This issue sounds like a duplicate of #2225

Must admit I have not cared much about optimizing for complete StackTrace-capture. Since performance hit is huge. I have made some changes for NLog 5.0 to improve the use of caller-member-attributes to provide callsite-details.

Think you should create your own custom layoutrenderer, that can extract and render the StackTrace that you have added as LogEvent Property. See also: https://github.com/NLog/NLog/wiki/How-to-write-a-custom-layout-renderer

**Comment by snakefoot** [comment]
Actually had the same thoughts when I implemented better support for caller-member-attributes.

Would like to have to have callsite-info from those locations that provide it automatically, but not give performance penalty (StackTrace-capture) for all those locations that doesn't give it upfront.

But put my thoughts on hold, because other dragons also needs to be attended.

**Comment by snakefoot** [comment]
@Rolemancer You are welcome to create a PR to improve the StackTraceUsage-logic, so one can force the different CallSite-LayoutRenderers to return StackTraceUsage.None. (So they will only render output if callsite-info has been provided manually/explicitly).

**Comment by snakefoot** [comment]
@Rolemancer Please review #3915 to see if it would solve your issue ?

**Comment by Rolemancer** [comment]
@snakefoot Looks cool! It should solve my problems! Thank you!

**Comment by snakefoot** [comment]
Closed after resolved by #3915

**Comment by snakefoot** [comment]
@Rolemancer NLog 4.7.1 has been released.
